src/hotspot/share/opto/memnode.cpp

```diff
@@ -2801,7 +2827,35 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
       val->in(MemNode::Address)->eqv_uncast(adr) &&
       val->in(MemNode::Memory )->eqv_uncast(mem) &&
       val->as_Load()->store_Opcode() == Opcode()) {
-    result = mem;
+        // Handle StoreVector with offsets and masks
+        // Ensure offsets match
+        if (is_StoreVectorScatter()) {
+          const Node* offsets = as_StoreVectorScatter()->in(StoreVectorScatterNode::Offsets);
+          if (val->is_LoadVectorGather() && offsets->eqv_uncast(val->as_LoadVectorGather()->in(LoadVectorGatherNode::Offsets))) {
+            result = mem;
+          }
+        // Ensure masks match
+        } else if (is_StoreVectorMasked()) {
+          const Node* mask = as_StoreVectorMasked()->in(StoreVectorMaskedNode::Mask);
+          if (val->is_LoadVectorMasked() && mask->eqv_uncast(val->as_LoadVectorMasked()->in(LoadVectorMaskedNode::Mask))) {
+            result = mem;
+          }
+        // Ensure offsets and masks match
+        } else if (is_StoreVectorScatterMasked()) {
+          const StoreVectorScatterMaskedNode* stv = as_StoreVectorScatterMasked();
+          const Node* offsets = stv->in(StoreVectorScatterMaskedNode::Offsets);
+          const Node* mask = stv->in(StoreVectorScatterMaskedNode::Mask);
+          if (val->is_LoadVectorGatherMasked()) {
+            const LoadVectorGatherMaskedNode* lvgm = val->as_LoadVectorGatherMasked();
+            if (offsets->eqv_uncast(lvgm->in(LoadVectorGatherMaskedNode::Offsets)) &&
+              mask->eqv_uncast(lvgm->in(LoadVectorGatherMaskedNode::Mask))) {
+              result = mem;
+            }
+          }
+        // Regular store (no offsets or mask)
+        } else {
+          result = mem;
+        }
```

Comment:
You already have the condition `val->as_Load()->store_Opcode() == Opcode()`.
So once you have `is_StoreVectorScatter()`, I think `val->is_LoadVectorGather()` is implied, no?

Oh wow. I think actually that this is another bug here: we only have
`virtual int store_Opcode() const { return Op_StoreVector; }` for `LoadVectorNode`, but not for all masked/gather/scatter vector nodes! I think that should be fixed.

That would also simplify your code.

Reply:
I’ve updated the `store_Opcode` methods. Actually I’ve used them to always avoid being the same as a Store opcode (I’ve used the Load own opcodes (a bit hacky) but that can be changed). See the comment below.